06-building-decays.md

```diff
@@ -0,0 +1,161 @@
+
+> ## Using the `DataOnDemand` selection {.callout}
+> As discussed previously, the `DataOnDemand` or `AutomaticData` selection builds objects from their TES location.
+> In Gaudi, the TES location of the output of an algorithm is generally determined as `Phys/ALGO_NAME/OBJECT_TYPE`, where `OBJECT_TYPE` refers to `Particles`, `Vertices`, etc.
+> For example, in our specific case, we could have used the `DataOnDemand` class with the `Phys/StdAllNoPIDsPions/Particles` and `Phys/StdAllLooseKaons/Particles` as inputs for pions and kaons, respectively:
+> 
+> ```python
+> from PhysSelPython.Wrappers import DataOnDemand
+> Pions = DataOnDemand('Phys/StdAllNoPIDsPions/Particles')
+> Kaons = DataOnDemand('Phys/StdAllLooseKaons/Particles')
+> ```
+>
+> In fact, if you access the [code](https://svnweb.cern.ch/trac/lhcb/browser/Stripping/trunk/Phys/CommonParticles/python/CommonParticles/StdAllNoPIDsPions.py) for `StdAllNoPIDsPions`, you can see that in it the algorithm for creating the particles is instantiated and the map of the `DataOnDemand` service is updated via the `updateDoD` function.
+>
+> Using the `DataOnDemand` service has the caveat that we need to hardcode the location of the particles, instead of loading a python class such as `StdAllNoPIDsPions`, and therefore must be used with care;
+> its usage is only necessary in a few non-trivial cases where we have no access to the algorithm that build the objects we need.
+
+
+Once we have the input pions and kaons, we can combine them to build a D0 by means of the `CombineParticles` algorithm.
+This algorithm performs the combinatorics for us according to a given decay descriptor and puts the resulting particle in the TES, allowing also to apply some cuts on them:
+
+ - `DaughtersCuts` is a dictionary that maps each child particle type to a LoKi particle functor that determines if that particular particle satisfies our selection criteria. Optionally, one can specify also a `Preambulo` property that allows us to make imports, preprocess functors, etc (more on this in the [LoKi functors](https://lhcb.github.io/first-analysis-steps/06-loki-functors.html) lesson). For example:
+
+```python
+d0_daughters = {'pi-': '(PT > 750*MeV) & (P > 4000*MeV) & (MIPCHI2DV(PRIMARY) > 4)',
+                'K+': '(PT > 750*MeV) & (P > 4000*MeV) & (MIPCHI2DV(PRIMARY) > 4)'}
+```
+
+ - `CombinationCut` is a particle array LoKi functor (note the `A` prefix, see more [here](https://twiki.cern.ch/twiki/bin/view/LHCb/LoKiHybridFilters#Particle_Array_Functors)) that is given the array of particles in a single combination (the *children*) as input (in our case a kaon and a pion). This cut is applied before the vertex fit so it is typically used to save CPU cycles by performing some sanity cuts such as `AMAXDOCA` or `ADAMASS` (do you know what they do?) before the CPU-consuming fit:
```

Comment:
Could move the “do you know what they do?” question to a `challenge` box at the end of the lesson.

> Earlier we used the `AMAXDOCA` and `ADAMASS` LoKi functors when creating D0 combinations. What physical quantities do these functors represent?